### DISCUSSION TITLE
Raspberry Pico, Thonny and serial output

### DISCUSSION THREAD

**Discussion by MatzeZett** [micropython/micropython#9619] [Using MicroPython]
Hello to everybody.

It seems to be quite a simple task but I cannot figure out why it doesn't work.

The Pico is connected to a Windows PC by an ordinary USB cable.
Programming works fine, so the cable must be intact.

What I want from the Pico is:
He should send me a string in every second.
What I write is:

import time
print('string\r\n')
time.sleep(1)

I can see the string in the console of Thonny.
Closing Thonny ( to release the COM-Port) and opening a terminal software also shows me this string every second.
But this only works with the serial Monitor of the Arduino-IDE and with Putty.

If i use hTerm or my own software ( which works perfectly with an Arduino nano in the same way) it shows nothing.

What I wonder is:
How does this micropython script know which baudrate it has to use?

Any help is veeery welcome.
Cheers,
Matt

**Comment by jimmo** [comment]
> print('string\r\n')

You shouldn't include the `\r` here -- the \r is added automatically in MicroPython's serial output code. Just write `print('string')` and if you want an _extra_ newline, then just `print('string\n')`.

> If i use hTerm or my own software ( which works perfectly with an Arduino nano in the same way) it shows nothing.

The fact that it works in the other programs is interesting. Does `mpremote` work?

> How does this micropython script know which baudrate it has to use?

There's no baud rate involved when talking over USB. It's a fake serial port (a CDC ACM device). When you set a baud rate, it's send to the Pico, but ignored by MicroPython.

This does matter when you're talking to a microcontroller that doesn't natively support USB and instead it's via a USB to UART converter on the board (e.g. an ESP32 and many older Arduino boards) -- the baud rate send to the USB/UART converter is then used to configure the UART that talks to the actual chip).

**Reply by MatzeZett** [reply]
Sorry but I cannot try mpremote. It is a Windows PC.

**Reply by jimmo** [reply]
`mpremote` works on Windows. Can be installed from pip.

**Reply by MatzeZett** [reply]
I am sorry, but I am not able to do such tasks. I am not a PC specialist nor a programmer.
I thought that it would be as simple as it looked.
Typing few lines and get an output to the USB port.
Obviously it is much more than that.

**Reply by jimmo** [reply]
The fact that it works in Thonny, Putty, and Arduino suggests that something is wrong in the way hTerm and your software are working. Are they opening the correct COM port?

**Reply by MatzeZett** [reply]
Yes they do.
And when I use the setup with the Arduino Nano instead of the Pico all software is doing well.

**Reply by MatzeZett** [reply]
Right now I am trying to use another Pico, that will take a few minutes.

**Reply by MatzeZett** [reply]
Nothing has changed when trying another Pico

**Reply by jimmo** [reply]
FWIW the main difference between an Arduino Nano and a Pico is that the Nano's USB is implemented via a FTDI USB/UART converter (the ATMega328 doesn't have USB support). So there must be something different about how hTerm is handling a FTDI vs a CDC ACM USB device.

**Reply by MatzeZett** [reply]
Ok, let's suggest that hTerm does this in a different way.
( There is nothing to be found on the internet about such a problem)

But why does my software not work, too?

The software is written in Visual Basic.
Source code is available, but very long.

**Reply by jimmo** [reply]
Have you tried removing the `\r` ? Can you access the REPL?

**Reply by MatzeZett** [reply]
I tried, yes.
I even tried to send a simple byte.
I tried it on a different Laptop - same as here.

Only one option is left for me:
Try it with C++
But setting up the IDE in Visual Studio is a thing that is far to much for me.

**Reply by MatzeZett** [reply]
I do not know how to handle the REPL ( I don't know what that exactly means )

**Reply by MatzeZett** [reply]
Maybe somebody has a Raspberry Pico and can try to use hTerm and tell me if you get a result?

**Reply by jimmo** [reply]
OK I tried it. Have you clicked the "DTR" button in hTerm? (Other programs enable this by default)

**Reply by jimmo** [reply]
hTerm is really not an ideal program to use with MicroPython because it's line oriented. It's also just really confusing. You want a "true" terminal (like Putty or mpremote). Again this DTR issue is because of the difference between FTDI (e.g. Arduino Nano) and an ACM device (Pico, and many others). You would see the same issue using hTerm with an Arduino Leonardo or many of the newer Arduino boards.

If you're looking for a beginner-friendly option for getting started with MicroPython, maybe try Thonny.

**Reply by MatzeZett** [reply]
jimmo - you're great !

DTR was the solution!
Now I get the desired Data on hTerm.

Yes, I already use Thonny. Great for beginners like me.
hTerm was just another one to double-check the output of the Pico.

OK, so now I have to find out why my own software does not receive data.

Thanks a lot
I will get back to you as soon as I can figure out something new.

Best wishes,
Matt

**Reply by MatzeZett** [reply]
In my C++ code I added DtrEnable = True

And now:    Everythings is running great !!!  Yippie

Many thanks again to you, without your help I wouldn't have been able to solve the problem.

Are you a model railroader?
